core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -140,6 +200,362 @@ protected[actions] trait PrimitiveActions {
+   * It parses the result and extracts the name of the next component action if any.
+   * It either invokes the desired component action or completes the composition invocation.
+   * It also checks the invocation counts against the limits.
+   *
+   * @param user the identity invoking the action
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeConductor(user: Identity, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId): Future[ActivationResponse] = {
+
+    if (session.accounting.conductors > 2 * actionSequenceLimit) {
+      // composition is too long
+      Future.successful(ActivationResponse.applicationError(compositionIsTooLong))
+    } else {
+      // inject state into payload if any
+      val params = session.state
+        .map(state => Some(JsObject(payload.getOrElse(JsObject()).fields ++ state.fields)))
+        .getOrElse(payload)
+
+      // invoke conductor action
+      session.accounting.conductors += 1
+      val activationResponse =
+        invokeSimpleAction(
+          user,
+          action = session.action,
+          payload = params,
+          waitForResponse = Some(session.action.limits.timeout.duration + 1.minute), // wait for result
+          cause = Some(session.activationId)) // cause is session id
+
+      waitForActivation(user, session, activationResponse).flatMap {
+        case Left(response) => // unsuccessful invocation, return error response
+          Future.successful(response)
+        case Right(activation) => // successful invocation
+          val result = activation.resultAsJson
+
+          // extract params from result, auto boxing result if not a dictionary
+          val params = result.fields.get(WhiskActivation.paramsField).map {
+            case obj: JsObject => obj
+            case value         => JsObject(WhiskActivation.valueField -> value)
+          }
+
+          // update session state, auto boxing state if not a dictionary
+          session.state = result.fields.get(WhiskActivation.stateField).map {
+            case obj: JsObject => obj
+            case value         => JsObject(WhiskActivation.stateField -> value)
+          }
+
+          // extract next action from result and invoke
+          result.fields.get(WhiskActivation.actionField) match {
+            case None =>
+              // no next action, end composition execution, return to caller
+              Future.successful(ActivationResponse(activation.response.statusCode, Some(params.getOrElse(result))))
+            case Some(next) =>
+              Try(next.convertTo[EntityPath]) match {
+                case Failure(t) =>
+                  // parsing failure
+                  Future.successful(ActivationResponse.applicationError(compositionComponentInvalid(next)))
+                case Success(next) =>
```

Comment:
could save a level of nesting by doing this:

```
               case Success(next) if (session.accounting.components >= actionSequenceLimit) =>
                    Future.successful(ActivationResponse.applicationError(compositionIsTooLong)) 
               case Success(next) => the rest
```

Reply:
yes!